File: lib/HabitCore/src/Stats.cpp
#include "habitcore/Stats.h"

namespace habitcore {

Date startOfWeek(const Date& date) {
  const int wd = static_cast<int>(date.weekday());  // Monday = 0
  return date.addDays(-wd);
}

StreakInfo computeStreak(const Schedule& schedule, const CompletionLog& log, const Date& today) {
  StreakInfo info;
  info.dueToday = schedule.isDue(today);
  info.doneToday = log.isDone(today);

  const std::vector<int32_t> doneDays = log.doneDays();
  info.completed = 0;
  for (int32_t dn : doneDays) {
    if (schedule.isDue(Date::fromDayNumber(dn))) ++info.completed;
  }

  if (schedule.isNever()) return info;

  const int32_t todayDn = today.toDayNumber();

  // Current streak: walk backwards over due days. Today is allowed grace: if it
  // is due but not done, we start counting from the previous due day instead of
  // treating the gap as broken. Any earlier missed due day stops the count.
  {
    int streak = 0;
    int32_t dn = todayDn;
    bool startedGrace = false;
    // Establish a lower bound so a habit with no history terminates quickly.
    int32_t floorDn = todayDn - 366 * 5;  // five years is plenty of headroom
    if (!doneDays.empty() && doneDays.front() < floorDn) floorDn = doneDays.front();

    while (dn >= floorDn) {
      const Date d = Date::fromDayNumber(dn);
      if (schedule.isDue(d)) {
        const bool done = log.isDone(dn);
        if (done) {
          ++streak;
        } else if (dn == todayDn && !startedGrace) {
          // Grace for an unfinished today: skip without breaking.
          startedGrace = true;
        } else {
          break;
        }
      }
      --dn;
    }
    info.current = streak;
  }

  // Longest streak: scan forward across the retained window counting the
  // longest run of consecutive completed due days.
  {
    int best = 0;
    int run = 0;
    if (!doneDays.empty()) {
      const int32_t first = doneDays.front();
      for (int32_t dn = first; dn <= todayDn; ++dn) {
        const Date d = Date::fromDayNumber(dn);
        if (!schedule.isDue(d)) continue;
        if (log.isDone(dn)) {
          ++run;
          if (run > best) best = run;
        } else {
          run = 0;
        }
      }
    }
    info.longest = best;
    if (info.current > info.longest) info.longest = info.current;
  }

  // Due days so far (from first activity to today), for a completion rate.
  {
    int due = 0;
    if (!doneDays.empty()) {
      for (int32_t dn = doneDays.front(); dn <= todayDn; ++dn) {
        if (schedule.isDue(Date::fromDayNumber(dn))) ++due;
      }
    }
    info.dueSoFar = due;
  }

  return info;
}

namespace {
// anchorDn is the first day the habit was ever logged. Scheduled days before it
// render Blank (the habit did not exist yet) rather than as missed.
CellState cellFor(const Schedule& schedule, const CompletionLog& log, const Date& date, int32_t todayDn,
                  int32_t anchorDn, bool hasAnchor) {
  const int32_t dn = date.toDayNumber();
  if (dn > todayDn) return CellState::Future;
  if (log.isDone(dn)) return CellState::Done;
  if (!schedule.isDue(date)) return CellState::NotDue;
  if (!hasAnchor || dn < anchorDn) return CellState::Blank;  // before the habit started
  return CellState::Due;                                     // scheduled, on/after start, not done
}
}  // namespace

YearGrid buildYearGrid(const Schedule& schedule, const CompletionLog& log, const Date& today, int weeks) {
  if (weeks < 1) weeks = 1;
  YearGrid grid;
  grid.weeks = weeks;
  grid.cells.resize(static_cast<size_t>(weeks) * 7);

  const int32_t todayDn = today.toDayNumber();
  int32_t anchorDn = 0;
  const bool hasAnchor = log.firstRecordedDay(anchorDn);
  // The right-most column is the week containing today; walk back `weeks-1`
  // whole weeks to find the first column's Monday.
  const Date thisMonday = startOfWeek(today);
  const Date firstMonday = thisMonday.addDays(-7 * (weeks - 1));

  for (int w = 0; w < weeks; ++w) {
    for (int wd = 0; wd < 7; ++wd) {
      const Date date = firstMonday.addDays(7 * w + wd);
      GridCell cell;
      cell.date = date;
      cell.state = cellFor(schedule, log, date, todayDn, anchorDn, hasAnchor);
      grid.cells[static_cast<size_t>(w) * 7 + wd] = cell;
    }
  }
  return grid;
}

WeekRow buildWeekRow(const Schedule& schedule, const CompletionLog& log, const Date& today) {
  WeekRow row;
  const int32_t todayDn = today.toDayNumber();
  int32_t anchorDn = 0;
  const bool hasAnchor = log.firstRecordedDay(anchorDn);
  const Date monday = startOfWeek(today);
  for (int wd = 0; wd < 7; ++wd) {
    const Date date = monday.addDays(wd);
    const CellState state = cellFor(schedule, log, date, todayDn, anchorDn, hasAnchor);
    row.days[wd] = state;
    if (state == CellState::Due || state == CellState::Done) ++row.dueCount;
    if (state == CellState::Done) ++row.doneCount;
  }
  return row;
}

}  // namespace habitcore

File: companion/habitink/stats.py
"""Streak and grid statistics, mirroring habitcore/Stats.cpp exactly.

Keeping the rules identical on both sides means the companion reports the same
numbers the device shows.
"""
from __future__ import annotations

import datetime as dt
from dataclasses import dataclass

from .model import CompletionLog, Schedule

ONE_DAY = dt.timedelta(days=1)


@dataclass
class StreakInfo:
    current: int = 0
    longest: int = 0
    completed: int = 0
    due_so_far: int = 0
    done_today: bool = False
    due_today: bool = False

    @property
    def rate(self) -> int:
        return (self.completed * 100) // self.due_so_far if self.due_so_far else 0


def start_of_week(date: dt.date) -> dt.date:
    return date - dt.timedelta(days=date.weekday())


def compute_streak(schedule: Schedule, log: CompletionLog, today: dt.date) -> StreakInfo:
    info = StreakInfo()
    info.due_today = schedule.is_due(today)
    info.done_today = log.is_done(today)

    done_dates = log.done_dates()
    info.completed = sum(1 for d in done_dates if schedule.is_due(d))

    if schedule.mask & 0x7F == 0:
        return info

    # Current streak with same-day grace.
    streak = 0
    day = today
    started_grace = False
    floor = today - dt.timedelta(days=366 * 5)
    if done_dates and done_dates[0] < floor:
        floor = done_dates[0]
    while day >= floor:
        if schedule.is_due(day):
            if log.is_done(day):
                streak += 1
            elif day == today and not started_grace:
                started_grace = True
            else:
                break
        day -= ONE_DAY
    info.current = streak

    # Longest run of consecutive completed due days.
    best = run = 0
    if done_dates:
        day = done_dates[0]
        while day <= today:
            if schedule.is_due(day):
                if log.is_done(day):
                    run += 1
                    best = max(best, run)
                else:
                    run = 0
            day += ONE_DAY
    info.longest = max(best, info.current)

    # Due days from first activity to today.
    due = 0
    if done_dates:
        day = done_dates[0]
        while day <= today:
            if schedule.is_due(day):
                due += 1
            day += ONE_DAY
    info.due_so_far = due
    return info


# Grid cell states mirror habitcore::CellState.
NOT_DUE = "not_due"
DUE = "due"
DONE = "done"
FUTURE = "future"
BLANK = "blank"


def cell_state(schedule: Schedule, log: CompletionLog, date: dt.date, today: dt.date,
               anchor: dt.date | None) -> str:
    if date > today:
        return FUTURE
    if log.is_done(date):
        return DONE
    if not schedule.is_due(date):
        return NOT_DUE
    if anchor is None or date < anchor:
        return BLANK
    return DUE


def build_year_grid(schedule: Schedule, log: CompletionLog, today: dt.date, weeks: int = 53) -> list[list[str]]:
    """Returns weeks columns, each a list of 7 cell states (Mon..Sun)."""
    anchor = log.first_recorded()
    this_monday = start_of_week(today)
    first_monday = this_monday - dt.timedelta(days=7 * (weeks - 1))
    grid = []
    for w in range(weeks):
        col = []
        for wd in range(7):
            date = first_monday + dt.timedelta(days=7 * w + wd)
            col.append(cell_state(schedule, log, date, today, anchor))
        grid.append(col)
    return grid


def month_days(year: int, month: int) -> list[dt.date]:
    first = dt.date(year, month, 1)
    if month == 12:
        nxt = dt.date(year + 1, 1, 1)
    else:
        nxt = dt.date(year, month + 1, 1)
    days = []
    d = first
    while d < nxt:
        days.append(d)
        d += ONE_DAY
    return days
